src/rcclReduceScatter.cpp

```diff
@@ -0,0 +1,313 @@
+                           rcclComm_t comm, hipStream_t stream) {
+    if ((RCCL_TRACE_RT & krccl_print_api) == krccl_print_api) {
+        int dev;
+        hipGetDevice(&dev);
+        fprintf(stderr,
+                "%s<<rccl-api:%s rccl-device:%d sendbuff:%p recvbuff:%p "
+                "recvcount:%d datatype:%s op:%s comm:%p stream:%p%s\n",
+                API_COLOR, __func__, dev, sendbuff, recvbuff, recvcount,
+                umap_datatype[datatype].c_str(), umap_red_op[op].c_str(), comm,
+                stream, API_COLOR_END);
+    }
+
+    //! Check if buffer pointers are not null
+    if (sendbuff == nullptr || recvbuff == nullptr) {
+        return rcclInvalidDevicePointer;
+    }
+
+    //! Check if data type of buffers is valid or not
+    if (datatype >= rccl_NUM_TYPES) {
+        return rcclInvalidType;
+    }
+
+    //! Check if op is valid or not
+    if (op >= rccl_NUM_OPS) {
+        return rcclInvalidOperation;
+    }
+
+    //! Get internal communicator from rcclComm_t
+    RcclComm_t *pcomm = comm;
+
+    //! Check if communicator is valid or number of elements is > 0
+    if (pcomm == nullptr || recvcount <= 0) {
+        return rcclInvalidArgument;
+    }
+
+    int rank = pcomm->rank_;
+    int num_gpus = pcomm->num_devices_;
+    hipEvent_t event = pcomm->event_;
+    int count = recvcount*num_gpus;
+
+    //! Get pointer to current barrier
+    int *this_time = &(pcomm->this_time_);
+
+    //! If same comm is used on a different stream, synchronize it with current
+    //! stream before launching op.
+    PreEnqueueEventRecord(pcomm, stream);
+
+    //! Get tracker to current gpu
+    RingNode_t *pcurr_track = pcomm->track_;
+
+    //! If the number of gpus equal to 1, do a simple memory copy
+    if (num_gpus == 1) {
+        switch (datatype) {
+        case rcclChar: {
+            hipMemcpyAsync(recvbuff, sendbuff, count * sizeof(char),
+                           hipMemcpyDeviceToDevice, stream);
+            break;
+        }
+        case rcclHalf: {
+            hipMemcpyAsync(recvbuff, sendbuff, count * sizeof(short),
+                           hipMemcpyDeviceToDevice, stream);
+            break;
+        }
+        case rcclInt:
+        case rcclFloat: {
+            hipMemcpyAsync(recvbuff, sendbuff, count * sizeof(int),
+                           hipMemcpyDeviceToDevice, stream);
+            break;
+        }
+        case rcclInt64:
+        case rcclUint64:
+        case rcclDouble: {
+            hipMemcpyAsync(recvbuff, sendbuff, count * sizeof(double),
+                           hipMemcpyDeviceToDevice, stream);
+            break;
+        }
+        default: { return rcclInvalidType; }
+        }
+
+        //! Track current stream so that op launched on different stream can be
+        //! synchronized with current stream
+        PostEnqueueEventRecord(pcomm, stream);
+        return rcclSuccess;
+    }
+
+    //! Check which op to launch
```

Comment:
There's quite a bit of code here required to expand out the 2 templated arguments (datatype / operation).
An alternative solution for dealing with this is to "de-onion" and deal with one template argument at a time:

For example:
```
foo_datatype(rcclDataType_t datatype, rcclRedOp_t op, ...)
{
   switch (datatype)
   {
     case rcclChar: return foo_op<signed char, rccl_char16_t>(op, COMMON_ARGS);
     case rcclHalf: return foo_op<__fp16, rccl_half8_t>(op, COMMON_ARGS);
     ...
}

template <typename T, rcclDataType_t datatype>
foo_op(rcclRedOp_t op, COMMON_ARGS)
{
  switch (op)
  {
  case rcclSum: return RcclInternalReduceScatter<T, datatype, rcclSum>(COMMON_ARGS);
  case rcclProd: return RcclInternalReduceScatter<T, datatype, rcclProd>(COMMON_ARGS);
  ...
}
```

This might provide more clarity to the code, and avoid any nasty one-off typos.  Also, it might make life easier if we have to add any additional template parameters.

Reply:
Are there any performance or resource usage benefits of making these changes?

Reply:
Unlikely.  Purely cosmetic.  Don't worry about it.